sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -497,23 +496,33 @@ public void setWindowedWrites(boolean windowedWrites) {
      * idempotent, as it may be executed multiple times in the case of failure or for redundancy. It
      * is a best practice to attempt to try to make this method atomic.
      *
+     * <p>Returns the set of temporary files generated. Callers must call
+     * {@link #removeTemporaryFiles(Set)} to cleanup these files.
+     *
      * @param writerResults the results of writes (FileResult).
      */
-    public void finalize(Iterable<FileResult<DestinationT>> writerResults) throws Exception {
-      // Collect names of temporary files and rename them.
+     public Set<ResourceId> finalize(
+         Iterable<FileResult<DestinationT>> writerResults) throws Exception {
+      // Collect names of temporary files and copies them.
       Map<ResourceId, ResourceId> outputFilenames = buildOutputFilenames(writerResults);
       copyToOutputFiles(outputFilenames);
+       return outputFilenames.keySet();
+    }
 
-      // Optionally remove temporary files.
-      // We remove the entire temporary directory, rather than specifically removing the files
-      // from writerResults, because writerResults includes only successfully completed bundles,
-      // and we'd like to clean up the failed ones too.
-      // Note that due to GCS eventual consistency, matching files in the temp directory is also
-      // currently non-perfect and may fail to delete some files.
-      //
-      // When windows or triggers are specified, files are generated incrementally so deleting
-      // the entire directory in finalize is incorrect.
-      removeTemporaryFiles(outputFilenames.keySet(), !windowedWrites);
+    /*
+     * Remove temporary files after finalization.
+     *
+     * <p>We remove the entire temporary directory, rather than specifically removing the files from
```

Comment:
Made this comment more precise. 

I'm not sure that the condition is unbounded pcollections. e.g. there's no reason a streaming or incremental runner couldn't be used on a bounded pcollection. The condition that ensures we won't be doing things incrementally is GlobalWindows with no trigger - which is reflected by !windowedWrites.